### PR TITLE
APIS-526/529: Filtering and pagination of developers list

### PR DESCRIPTION
**Pull request by rm-hull**
Filtering by API (changes also in third-party-developer, third-party-application - see associated PRs) - API subscriptions are fused into the `ApplicationResponse`, and the controller filters only those devs who are collaborators on applications that have subscribed to the nominated API.

Pagination - rather than using client-side filtering with dataTables (the list could be many thousands of items), it is implemented serverside

### CODE REVIEW

File: app/connectors/ApiDefinitionConnector.scala [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import scala.concurrent.Future
+
+trait ApiDefinitionConnector {
+  val serviceBaseUrl: String
+  val http: HttpGet
+
+  def fetchAll()(implicit hc: HeaderCarrier): Future[Seq[APIDefinition]] = {
+    http.GET[Seq[APIDefinition]](s"$serviceBaseUrl/api-definition")
+      .recover {
+        case e: Upstream5xxResponse => throw new FetchApplicationsFailed
```
Comment (hughfdjackson) [inline app/connectors/ApiDefinitionConnector.scala]: Should this be `FetchAPIDefinitionsFailed`?

Comment (rm-hull) [inline app/connectors/ApiDefinitionConnector.scala]: Yes, you're right... have altered and added some tests + fixed the failing acceptance tests

File: app/connectors/ApiDefinitionConnector.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+trait ApiDefinitionConnector {
+  val serviceBaseUrl: String
+  val http: HttpGet
+
+  def fetchAll()(implicit hc: HeaderCarrier): Future[Seq[APIDefinition]] = {
+    http.GET[Seq[APIDefinition]](s"$serviceBaseUrl/api-definition")
+      .recover {
+        case e: Upstream5xxResponse => throw new FetchApiDefinitionFailed
```
Comment (hughfdjackson) [inline app/connectors/ApiDefinitionConnector.scala]: should be `FetchApiDefinitionsFailed` (in the plural)?

File: app/connectors/ApiDefinitionConnector.scala (hunk 1 above)
Comment (hughfdjackson) [inline app/connectors/ApiDefinitionConnector.scala]: hah - apparently I made the same comment twice :p

**Comment by hughfdjackson** [comment]
@rm-hull: I'm not convinced that a frontend should have the logic in to: 

* Fetch all applications
* Fetch all APIs
* Filter applications by subscriptions to a particular API

I think this should rightly live in third-party-application - as a filter on a fetch all applications endpoint.  e.g.:

```
GET third-party-application.service/applications?subscribesTo=<api-name>
Authorization: Bearer <gatekeeper BT>
```

**Comment by rm-hull** [comment]
@hughfdjackson: Ok, sure I can push the filtering back into third-party-application, but the front-end still needs to fetch the API definitions list for displaying the dropdown.

**Comment by hughfdjackson** [comment]
@rm-hull: that sounds reasonable to me.

File: app/connectors/ApplicationConnector.scala [hunk 3]
```diff
@@ -62,4 +62,11 @@ trait ApplicationConnector {
   def fetchApplication(applicationId: String)(implicit hc: HeaderCarrier): Future[ApplicationWithHistory] = {
     http.GET[ApplicationWithHistory](s"$applicationBaseUrl/gatekeeper/application/$applicationId")
   }
+
+  def fetchAllApplicationsBySubscription(apiContext: Option[String])(implicit hc: HeaderCarrier): Future[Seq[ApplicationResponse]] = {
+    http.GET[Seq[ApplicationResponse]](s"$applicationBaseUrl/application?subscribesTo=${apiContext.getOrElse("")}")
+      .recover {
+        case e: Upstream5xxResponse => throw new FetchApplicationsFailed
+      }
+  }
```
Comment (hughfdjackson) [inline app/connectors/ApplicationConnector.scala]: I would not assume that an application should treat: 

```
/path?query=
```

The same as 
```
/path
```
Which seems to be your intention here?  I can see the motivation for it below, because it allows convenient filtering. 

The Connectors should be fairly dumb/agnostic of the service use-cases below, though, so I'd suggest two different methods: 

```
def fetchAllApplicationsBySubscription(apiContext: String)(implicit hc: HeaderCarrier) = ???

def fetchAllApplications()(implicit hc: HeaderCarrier) = ??? 
```

File: app/controllers/DevelopersController.scala [hunk 4]
```diff
@@ -16,29 +16,65 @@
...
 trait DevelopersController extends FrontendController with GatekeeperAuthWrapper {
 
   val developerConnector: DeveloperConnector
+  val apiDefinitionConnector: ApiDefinitionConnector
+  val applicationConnector: ApplicationConnector
+
+  private def innerJoin(devs: Seq[User], apps: Seq[ApplicationResponse]): Seq[User] = {
+    val collaborators = apps.flatMap(_.collaborators).map(_.emailAddress).toSet
+    devs.filter(u => collaborators.contains(u.email))
+  }
```
Comment (hughfdjackson) [inline app/controllers/DevelopersController.scala]: Fetching all developer emails + names for all apps that subscribe to API Y (or all applications, for that matter) should be a service concern IMO.  

Controllers should ideally be as smarts-free as possible, except for where that's concerned with dealing with HTTP/view-rendering specific logic.

File: app/model/APIDefinition.scala [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                          versions: Seq[APIVersion],
+                          requiresTrust: Option[Boolean]) {
+
+  require(serviceName.nonEmpty, s"serviceName is required")
+  require(serviceBaseUrl.nonEmpty, s"serviceBaseUrl is required")
+  require(name.nonEmpty, s"name is required")
+  require(context.nonEmpty, s"context is required")
+  require(description.nonEmpty, s"description is required")
+  require(versions.nonEmpty, s"at least one version is required")
+  require(uniqueVersions, s"version numbers must be unique")
```
Comment (hughfdjackson) [inline app/model/APIDefinition.scala]: Probably not necessary to enforce requirements about data that's owned by a different service?

File: test/unit/connectors/ApplicationConnectorSpec.scala [hunk 6]
```diff
@@ -109,10 +109,31 @@ class ApplicationConnectorSpec extends UnitSpec with Matchers with ScalaFutures
...
+  "fetchAllApplicationsBySubscription" should {
+    "retrieve all applications subscribed to a specific API" in new Setup {
+      stubFor(get(urlEqualTo(s"/application?subscribesTo=some-context")).willReturn(aResponse().withStatus(200)
+        .withBody("[]")))
+
+      val result = await(connector.fetchAllApplicationsBySubscription(Some("some-context")))
+
+      verify(1, getRequestedFor(urlPathEqualTo("/application?subscribesTo=some-context"))
+        .withHeader("Authorization", equalTo(authToken)))
+    }
```
Comment (hughfdjackson) [inline test/unit/connectors/ApplicationConnectorSpec.scala]: There appears to be a missing test cases for the `None` context - although as noted above, I think that would be better as another method on the connector.

File: app/model/APIDefinition.scala (hunk 5 above)
Comment (rm-hull) [inline app/model/APIDefinition.scala]: Fixed

**Comment by antroy** [comment]
Merging with the outstanding "Logic into a Service" changes omitted.
